mirrord/sip/src/lib.rs

```diff
@@ -0,0 +1,407 @@
+                            SipError::UnsupportedFileFormat(
+                                "MachO 32 file parsing failed".to_string(),
+                            )
+                        })?;
+                    if header.cputype(Endianness::default()) == CPU_TYPE_X86_64 {
+                        Ok(Self::new(0, bytes.len()))
+                    } else {
+                        Err(SipError::NoX64Arch)
+                    }
+                }
+                FileKind::MachO64 => {
+                    let header: &MachHeader64<Endianness> =
+                        MachHeader::parse(bytes, 0).map_err(|_| {
+                            SipError::UnsupportedFileFormat(
+                                "MachO 64 file parsing failed".to_string(),
+                            )
+                        })?;
+                    if header.cputype(Endianness::default()) == CPU_TYPE_X86_64 {
+                        Ok(Self::new(0, bytes.len()))
+                    } else {
+                        Err(SipError::NoX64Arch)
+                    }
+                }
+                FileKind::MachOFat32 => FatHeader::parse_arch32(bytes)
+                    .map_err(|_| SipError::UnsupportedFileFormat("FatMach-O 32-bit".to_string()))?
+                    .iter()
+                    .find(is_fat_x64_arch)
+                    .map(|arch| Self::new(arch.offset() as usize, arch.size() as usize))
+                    .ok_or(SipError::NoX64Arch),
+                FileKind::MachOFat64 => FatHeader::parse_arch64(bytes)
+                    .map_err(|_| SipError::UnsupportedFileFormat("Mach-O 32-bit".to_string()))?
+                    .iter()
+                    .find(is_fat_x64_arch)
+                    .map(|arch| Self::new(arch.offset() as usize, arch.size() as usize))
+                    .ok_or(SipError::NoX64Arch),
+                other => Err(SipError::UnsupportedFileFormat(format!("{:?}", other))),
+            }
+        }
+    }
+
+    /// Patches a binary to disable SIP.
+    /// Right now it extracts x64 binary from fat/MachO binary and patches it.
+    fn patch_binary<P: AsRef<Path>, K: AsRef<Path>>(path: P, output: K) -> Result<()> {
+        let data = std::fs::read(path.as_ref())?;
+        let binary_info = BinaryInfo::from_object_bytes(&data)?;
+
+        let x64_binary = &data[binary_info.offset..binary_info.offset + binary_info.size];
+        std::fs::write(output.as_ref(), x64_binary)?;
+        std::fs::set_permissions(output.as_ref(), Permissions::from_mode(0o755))?;
+        codesign::sign(output)
+    }
+
+    /// Creates a new file at `patched_path` which has the same contents as `original_path` except
+    /// for the shebang which is `new_shebang`.
+    fn patch_script<P: AsRef<Path>, K: AsRef<Path>>(
+        original_path: P,
+        patched_path: K,
+        new_shebang: &str,
+    ) -> Result<()> {
+        if let Some(original_shebang) = read_shebang_from_file(&original_path)? {
```

Comment:
I think a better pattern here would be `.map`? not sure.

Reply:
LIke this?